src/prefect/server/utilities/leasing.py

```diff
@@ -66,6 +66,141 @@ async def revoke_lease(self, lease_id: UUID) -> None:
+    """
+    if lease_storage is None:
+        # Local import to avoid circular dependency during module import
+        from prefect.server.concurrency.lease_storage import (
+            get_concurrency_lease_storage,
+        )
+
+        lease_storage = get_concurrency_lease_storage()
+
+    # Convert tags to V2 names to check relevant limits
+    v2_names = [f"tag:{tag}" for tag in tags]
+
+    async with db.session_context() as session:
+        limit_ids: list[UUID] = []
+        for v2_name in v2_names:
+            model = await cl_v2_models.read_concurrency_limit(
+                session=session, name=v2_name
+            )
+            if model:
+                limit_ids.append(model.id)
+
+    if not limit_ids:
+        return None
+
+    # Use storage-provided holder search (now guaranteed by protocol)
+    desired = {"type": "task_run", "id": str(task_run_id)}
+    for limit_id in limit_ids:
+        holders = await lease_storage.list_holders_for_limit(limit_id)
+        # holders may be shape 1) {"type","id"} or 2) {"holder": {...}, "slots": N}
+        for h in holders:
+            payload = h.get("holder", h)
+            if payload == desired:
+                # Find the lease id by scanning active leases for this limit
+                for lid in await lease_storage.read_active_lease_ids(limit=1000):
+                    lease = await lease_storage.read_lease(lid)
+                    if lease and limit_id in lease.resource_ids:
+                        inner = (
+                            getattr(lease.metadata, "holder", None)
+                            if lease.metadata
+                            else None
+                        )
+                        if inner is not None and hasattr(inner, "model_dump"):
+                            inner = inner.model_dump(mode="json")  # type: ignore[attr-defined]
+                        if isinstance(inner, dict) and inner == desired:
+                            return lid
+
+    # Scan active leases if not found via holder search
+    active_leases = await lease_storage.read_active_lease_ids(limit=1000)
+
+    for lease_id in active_leases:
+        lease = await lease_storage.read_lease(lease_id)
+        if lease and lease.metadata and getattr(lease.metadata, "holder", None):
+            holder = lease.metadata.holder
+            if holder is not None and hasattr(holder, "model_dump"):
+                holder = holder.model_dump(mode="json")  # type: ignore[attr-defined]
+            if (
+                isinstance(holder, dict)
+                and holder.get("type") == "task_run"
+                and holder.get("id") == str(task_run_id)
+            ):
+                # Check if this lease is for one of the requested limits
+                if any(lid in lease.resource_ids for lid in limit_ids):
+                    return lease_id
+
+    return None
+
     async def read_expired_lease_ids(self, limit: int = 100) -> list[UUID]:
```

Comment:
The formatting is off here. This should still be on the `LeaseStorage` protocol.